SignalServiceKit/src/Messages/Interactions/OWSContactShare.m

```diff
@@ -262,34 +237,89 @@ - (NSString *)debugDescription
 
 #pragma mark -
 
-@interface OWSContact ()
+@implementation OWSContactName
+
+- (NSString *)logDescription
+{
+    NSMutableString *result = [NSMutableString new];
+    [result appendString:@"["];
+
+    if (self.givenName.length > 0) {
+        [result appendFormat:@"givenName: %@, ", self.givenName];
+    }
+    if (self.familyName.length > 0) {
+        [result appendFormat:@"familyName: %@, ", self.familyName];
+    }
+    if (self.middleName.length > 0) {
+        [result appendFormat:@"middleName: %@, ", self.middleName];
+    }
+    if (self.namePrefix.length > 0) {
+        [result appendFormat:@"namePrefix: %@, ", self.namePrefix];
+    }
+    if (self.nameSuffix.length > 0) {
+        [result appendFormat:@"nameSuffix: %@, ", self.nameSuffix];
+    }
+    if (self.displayName.length > 0) {
+        [result appendFormat:@"displayName: %@, ", self.displayName];
+    }
+
+    [result appendString:@"]"];
+    return result;
+}
 
-@property (nonatomic, nullable) NSString *givenName;
-@property (nonatomic, nullable) NSString *familyName;
-@property (nonatomic, nullable) NSString *nameSuffix;
-@property (nonatomic, nullable) NSString *namePrefix;
-@property (nonatomic, nullable) NSString *middleName;
-@property (nonatomic, nullable) NSString *organizationName;
-@property (nonatomic) NSString *displayName;
+- (NSString *)displayName
+{
+    [self ensureDisplayName];
 
-@property (nonatomic) NSArray<OWSContactPhoneNumber *> *phoneNumbers;
-@property (nonatomic) NSArray<OWSContactEmail *> *emails;
-@property (nonatomic) NSArray<OWSContactAddress *> *addresses;
+    if (_displayName.length < 1) {
+        OWSProdLogAndFail(@"%@ could not derive a valid display name.", self.logTag);
+        return NSLocalizedString(@"CONTACT_WITHOUT_NAME", @"Indicates that a contact has no name.");
+    }
+    return _displayName;
+}
 
-@property (nonatomic, nullable) NSString *avatarAttachmentId;
-@property (nonatomic) BOOL isProfileAvatar;
+- (void)ensureDisplayName
+{
+    if (_displayName.length < 1) {
+        CNContact *_Nullable systemContact = [OWSContactConversion systemContactForContactName:self];
+        _displayName = [CNContactFormatter stringFromContact:systemContact style:CNContactFormatterStyleFullName];
+    }
+    if (_displayName.length < 1) {
+        // Fall back to using the organization name.
+        _displayName = self.organizationName;
+    }
+}
+
+- (void)updateDisplayName
+{
+    _displayName = nil;
+
+    [self ensureDisplayName];
+}
+
++ (OWSContactName *)emptyName
```

Comment:
Why does this method exist? Why not inline `[OWContactName new]` instead?

Reply:
Removed.